File: maps/src/main/java/it/localhostsoftware/maps/model/PolygonOptions.java
package it.localhostsoftware.maps.model;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.gms.common.GoogleApiAvailability;
import com.huawei.hms.api.HuaweiApiAvailability;

import java.util.List;

import it.localhostsoftware.maps.google.model.GooglePolygonOptions;
import it.localhostsoftware.maps.huawei.model.HuaweiPolygonOptions;

public abstract class PolygonOptions<PO> {
    public static PolygonOptions<?> getInstance(Context context) {
        if (GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(context) == com.google.android.gms.common.ConnectionResult.SUCCESS)
            return new GooglePolygonOptions(new com.google.android.gms.maps.model.PolygonOptions());
        else if (HuaweiApiAvailability.getInstance().isHuaweiMobileServicesAvailable(context) == com.huawei.hms.api.ConnectionResult.SUCCESS)
            return new HuaweiPolygonOptions(new com.huawei.hms.maps.model.PolygonOptions());
        else throw new IllegalStateException();
    }

    private final PO po;

    public PolygonOptions(PO po) {
        this.po = po;
    }

    public PO getPolygonOptions() {
        return po;
    }

    @NonNull
    public abstract PolygonOptions<?> add(@NonNull LatLng<?> point);

    @NonNull
    public abstract PolygonOptions<?> add(@NonNull LatLng<?>... points);

    @NonNull
    public abstract PolygonOptions<?> addAll(@NonNull Iterable<LatLng<?>> points);

    @NonNull
    public abstract PolygonOptions<?> addHole(@NonNull Iterable<LatLng<?>> points);

    @NonNull
    public abstract PolygonOptions<?> strokeWidth(float width);

    @NonNull
    public abstract PolygonOptions<?> strokeColor(int color);

    @NonNull
    public abstract PolygonOptions<?> strokeJointType(int jointType);

    @NonNull
    public abstract PolygonOptions<?> strokePattern(@Nullable List<PatternItem<?>> pattern);

    @NonNull
    public abstract PolygonOptions<?> fillColor(int color);

    @NonNull
    public abstract PolygonOptions<?> zIndex(float zIndex);

    @NonNull
    public abstract PolygonOptions<?> visible(boolean visible);

    @NonNull
    public abstract PolygonOptions<?> geodesic(boolean geodesic);

    @NonNull
    public abstract PolygonOptions<?> clickable(boolean clickable);

    @NonNull
    public abstract List<LatLng<?>> getPoints();

    @NonNull
    public abstract List<List<LatLng<?>>> getHoles();

    public abstract float getStrokeWidth();

    public abstract int getStrokeColor();

    public abstract int getStrokeJointType();

    @Nullable
    public abstract List<PatternItem<?>> getStrokePattern();

    public abstract int getFillColor();

    public abstract float getZIndex();

    public abstract boolean isVisible();

    public abstract boolean isGeodesic();

    public abstract boolean isClickable();
}

File: maps/src/main/java/it/localhostsoftware/maps/model/PolygonOptions.kt
package it.localhostsoftware.maps.model

import android.content.Context
import it.localhostsoftware.maps.MobileServices
import it.localhostsoftware.maps.getMobileServices
import it.localhostsoftware.maps.google.model.GooglePolygonOptions
import it.localhostsoftware.maps.huawei.model.HuaweiPolygonOptions

abstract class PolygonOptions<PO, PI : PatternItem<*>, LL : LatLng<*>>(val po: PO) {
    companion object {
        fun getInstance(c: Context) =
            when (c.getMobileServices()) {
                MobileServices.GOOGLE -> GooglePolygonOptions(com.google.android.gms.maps.model.PolygonOptions())
                MobileServices.HUAWEI -> HuaweiPolygonOptions(com.huawei.hms.maps.model.PolygonOptions())
                else -> throw IllegalStateException()
            }
    }

    abstract fun add(vararg points: LL): PolygonOptions<*, *, *>
    abstract fun addHole(points: Iterable<LL>): PolygonOptions<*, *, *>
    abstract val points: List<LL>
    abstract val holes: List<List<LL>>
    abstract var strokeWidth: Float
    abstract var strokeColor: Int
    abstract var strokeJointType: Int
    abstract var strokePattern: List<PI>?
    abstract var fillColor: Int
    abstract var zIndex: Float
    abstract var isVisible: Boolean
    abstract var isGeodesic: Boolean
    abstract var isClickable: Boolean
    fun addAll(var1: Iterable<LL>): PolygonOptions<*, *, *> {
        var1.forEach { add(it) }
        return this
    }
}